### PR TITLE
Report script loads from eval when debugging

### PR DESCRIPTION
**Pull request by kfarnung**
Added code to report the load of scripts on eval in order to support
stepping into the eval code.

### CODE REVIEW

File: lib/Jsrt/JsrtDebugManager.cpp [hunk 1]
```diff
@@ -146,14 +146,26 @@ HRESULT JsrtDebugManager::DbgRegisterFunction(Js::ScriptContext* scriptContext,
 
     if (!utf8SourceInfo->GetIsLibraryCode() && !utf8SourceInfo->HasDebugDocument())
     {
+        JsrtDebugEventObject debugEventObject(scriptContext);
+        Js::DynamicObject* eventDataObject = debugEventObject.GetEventDataObject();
+
+        JsrtDebugUtils::AddFileNameOrScriptTypeToObject(eventDataObject, utf8SourceInfo);
+        JsrtDebugUtils::AddLineCountToObject(eventDataObject, utf8SourceInfo);
+        JsrtDebugUtils::AddPropertyToObject(eventDataObject, JsrtDebugPropertyId::sourceLength, utf8SourceInfo->GetCchLength(), utf8SourceInfo->GetScriptContext());
+
```
Comment (akroshg) [inline lib/Jsrt/JsrtDebugManager.cpp]: I see this operations in order and also appear on the other functions in this file. Could you refactor to one and call in the other functions as well?

File: lib/Runtime/Debug/DebugContext.h [hunk 2]
```diff
@@ -60,11 +60,14 @@ namespace Js
 
         HostDebugContext * GetHostDebugContext() const { return hostDebugContext; }
 
+        bool GetIsReparsingSource() { return this->isReparsingSource; }
```
Comment (agarwal-sandeep) [inline lib/Runtime/Debug/DebugContext.h]: use const

File: lib/Runtime/Debug/DebugContext.cpp [hunk 3]
```diff
@@ -141,6 +142,23 @@ namespace Js
 
     HRESULT DebugContext::RundownSourcesAndReparse(bool shouldPerformSourceRundown, bool shouldReparseFunctions)
     {
+        struct AutoRestore
```
Comment (agarwal-sandeep) [inline lib/Runtime/Debug/DebugContext.cpp]: >AutoRestore [](start = 15, length = 11)

Rename to AutoRestoreIsReparsingSource

File: lib/Runtime/Debug/DebugContext.cpp [hunk 4]
```diff
@@ -141,6 +142,23 @@ namespace Js
...
+        {
+            AutoRestore(DebugContext* debugContext)
+                :debugContext(debugContext)
+            {
+                this->debugContext->isReparsingSource = true;
```
Comment (agarwal-sandeep) [inline lib/Runtime/Debug/DebugContext.cpp]: > this->debugContext->isReparsingSource = true; [](start = 15, length = 46)

We should restrict this to only when shouldReparseFunctions is true. Not needed during PerformSourceRundown

File: lib/Jsrt/JsrtDebugManager.cpp [hunk 5]
```diff
@@ -154,6 +154,21 @@ HRESULT JsrtDebugManager::DbgRegisterFunction(Js::ScriptContext* scriptContext,
         {
             utf8SourceInfo->SetDebugDocument(debugDocument);
         }
+
+        if (this->debugEventCallback != nullptr && !scriptContext->GetDebugContext()->GetIsReparsingSource())
+        {
+            JsrtDebugEventObject debugEventObject(scriptContext);
+            Js::DynamicObject* eventDataObject = debugEventObject.GetEventDataObject();
+            JsrtDebugUtils::AddSourceMetadataToObject(eventDataObject, utf8SourceInfo);
```
Comment (agarwal-sandeep) [inline lib/Jsrt/JsrtDebugManager.cpp]: >AddSourceMetadataToObject [](start = 28, length = 25)

When ever we use AddSourceMetadataToObject we also add Scriptid. Move the HasDebugDocument check and AddScriptIdToObject inside AddSourceMetadataToObject

File: lib/Jsrt/JsrtDebugManager.cpp [hunk 6]
```diff
@@ -154,6 +154,21 @@ HRESULT JsrtDebugManager::DbgRegisterFunction(Js::ScriptContext* scriptContext,
         {
             utf8SourceInfo->SetDebugDocument(debugDocument);
         }
+
+        if (this->debugEventCallback != nullptr && !scriptContext->GetDebugContext()->GetIsReparsingSource())
```
Comment (agarwal-sandeep) [inline lib/Jsrt/JsrtDebugManager.cpp]: >!scriptContext->GetDebugContext()->GetIsReparsingSource()) [](start = 51, length = 58)

Add a comment here explaining why this check

File: lib/Jsrt/JsrtDebugManager.cpp (hunk 5 above)
Comment (kfarnung) [inline lib/Jsrt/JsrtDebugManager.cpp]: This is the case here, but not elsewhere in the file.  That's why I kept is separate.

File: lib/Jsrt/JsrtDebugManager.cpp [hunk 7]
```diff
@@ -154,6 +154,18 @@ HRESULT JsrtDebugManager::DbgRegisterFunction(Js::ScriptContext* scriptContext,
         {
             utf8SourceInfo->SetDebugDocument(debugDocument);
         }
+
+        // Only raise an event in the case where we have a registered callback and we are not
+        // currently in the middle of a source reparse (e.g. attaching/detaching).
```
Comment (agarwal-sandeep) [inline lib/Jsrt/JsrtDebugManager.cpp]: >middle of a source reparse (e.g. attaching/detaching). [](start = 28, length = 54)

Please explain why we are adding isParsingSource check, will help somebody understand the code/logic when we are not around.

**Comment by agarwal-sandeep** [comment]
:shipit:

**Pull request merged**